lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,783 @@
+/// capture pointers back into the parse tree, so the parse tree data structure
+/// may <em>not</em> be changed between the construction of the AST and all of
+/// its uses.
+///
+/// The AST captures a structured view of the program.  The program is a list of
+/// units.  Function like units will contain lists of evaluations.  Evaluations
+/// are either statements or constructs, where a construct contains a list of
+/// evaluations.  The resulting AST structure can then be used to create FIR.
+
+namespace Br = Fortran::lower;
+namespace Co = Fortran::common;
+namespace L = llvm;
+namespace Pa = Fortran::parser;
+
+using namespace Fortran;
+using namespace Br;
+
+namespace {
+
+/// The instantiation of a parse tree visitor (Pre and Post) is extremely
+/// expensive in terms of compile and link time, so one goal here is to limit
+/// the bridge to one such instantiation.
+class ASTBuilder {
+public:
+  ASTBuilder() {
+    pgm = new AST::Program;
+    parents.push_back(pgm);
+  }
+
+  /// Get the result
+  AST::Program *result() { return pgm; }
+
+  template <typename A>
+  constexpr bool Pre(const A &) {
+    return true;
+  }
+  template <typename A>
+  constexpr void Post(const A &) {}
+
+  // Module like
+
+  bool Pre(const Pa::Module &x) { return enterModule(x); }
+  bool Pre(const Pa::Submodule &x) { return enterModule(x); }
+
+  void Post(const Pa::Module &) { exitModule(); }
+  void Post(const Pa::Submodule &) { exitModule(); }
+
+  // Function like
+
+  bool Pre(const Pa::MainProgram &x) { return enterFunc(x); }
+  bool Pre(const Pa::FunctionSubprogram &x) { return enterFunc(x); }
+  bool Pre(const Pa::SubroutineSubprogram &x) { return enterFunc(x); }
+  bool Pre(const Pa::SeparateModuleSubprogram &x) { return enterFunc(x); }
+
+  void Post(const Pa::MainProgram &) { exitFunc(); }
+  void Post(const Pa::FunctionSubprogram &) { exitFunc(); }
+  void Post(const Pa::SubroutineSubprogram &) { exitFunc(); }
+  void Post(const Pa::SeparateModuleSubprogram &) { exitFunc(); }
+
+  // Block data
+
+  void Post(const Pa::BlockData &x) {
+    AST::BlockDataUnit unit{x, parents.back()};
+    addUnit(unit);
+  }
+
+  //
+  // Action statements
+  //
+
+  void Post(const Pa::Statement<Pa::ActionStmt> &s) {
+    addEval(makeEvalAction(s));
+  }
+  void Post(const Pa::UnlabeledStatement<Pa::ActionStmt> &s) {
+    addEval(makeEvalAction(s));
+  }
+
+  //
+  // Non-executable statements
+  //
+
+  void Post(const Pa::Statement<Co::Indirection<Pa::FormatStmt>> &s) {
+    addEval(makeEvalIndirect(s));
+  }
+  void Post(const Pa::Statement<Co::Indirection<Pa::EntryStmt>> &s) {
+    addEval(makeEvalIndirect(s));
+  }
+  void Post(const Pa::Statement<Co::Indirection<Pa::DataStmt>> &s) {
+    addEval(makeEvalIndirect(s));
+  }
+  void Post(const Pa::Statement<Co::Indirection<Pa::NamelistStmt>> &s) {
+    addEval(makeEvalIndirect(s));
+  }
+
+  //
+  // Construct statements
+  //
+
+  void Post(const Pa::Statement<Pa::AssociateStmt> &s) {
```

Comment:
I agree this could be condensed, but this might make it harder to work with and right now Val is also working on this file so I do not want to lock him up if he needs to modify these.

Reply:
After discussing with Val, I have condensed this as suggested.